components/sup/src/manager/service/pipe_hook_client.rs

```diff
@@ -0,0 +1,319 @@
+               OpenOptions},
+          io::{self,
+               prelude::*,
+               BufRead,
+               BufReader,
+               Read},
+          iter::once,
+          os::windows::{ffi::OsStrExt,
+                        fs::*,
+                        io::*},
+          path::PathBuf,
+          thread,
+          time::{Duration,
+                 Instant}};
+use uuid::Uuid;
+use winapi::um::{namedpipeapi,
+                 processthreadsapi,
+                 winbase};
+
+static LOGKEY: &str = "HK";
+const PIPE_SERVICE_ENVVAR: &str = "HAB_HOOK_PIPE_SCRIPT";
+const START_PIPE_TIMEOUT_SECONDS: u64 = 5;
+const SIGNAL_EXEC_HOOK: [u8; 1] = [1];
+const SIGNAL_QUIT: [u8; 1] = [0];
+
+#[derive(Debug)]
+pub struct PipeHookClient {
+    hook_name:       String,
+    hook_path:       PathBuf,
+    pipe_name:       String,
+    stdout_log_file: PathBuf,
+    stderr_log_file: PathBuf,
+}
+
+impl PipeHookClient {
+    pub fn new(hook_name: String,
+               hook_path: PathBuf,
+               stdout_log_file: PathBuf,
+               stderr_log_file: PathBuf)
+               -> Self {
+        let pipe_name = format!("{}-hook-{}", hook_name, Uuid::new_v4());
+        Self { hook_name,
+               hook_path,
+               pipe_name,
+               stdout_log_file,
+               stderr_log_file }
+    }
+
+    pub fn exec_hook<T>(&self,
+                        service_group: &str,
+                        pkg: &Pkg,
+                        svc_encrypted_password: Option<T>)
+                        -> Result<u32>
+        where T: ToString
+    {
+        // Always check to see if the powershell named pipe server is running.
+        // This should only be false the first time this function is called
+        // or if someone explicitly terminates the powershell process
+        if self.pipe_exists().is_err() {
+            debug!("Initiating Powershell named pipe server for {} {} hook",
+                   service_group, self.hook_name);
+            self.init_server(service_group, pkg, svc_encrypted_password)?;
+        }
+
+        let (mut pipe, poll) = self.connect()?;
+        debug!("connected to {} {} hook pipe",
+               service_group, self.hook_name);
+
+        // The powershell server takes a single byte as input which will be either
+        // 0 to shut down (see drop below) or 1 to run the hook
+        self.pipe_ready(&poll, Ready::writable())?;
+        pipe.write_all(&SIGNAL_EXEC_HOOK)?;
+
+        // Now we wait for the hook to run and the powershell service to
+        // send back the hook's exit code over the pipe
+        self.pipe_ready(&poll, Ready::readable())?;
+        let mut exit_buf = [0; 4];
+        pipe.read_exact(&mut exit_buf)?;
+        Ok(u32::from_ne_bytes(exit_buf))
+    }
+
+    fn pipe_ready(&self, poll: &Poll, readiness: Ready) -> Result<()> {
+        let mut events = Events::with_capacity(1024);
+        loop {
+            poll.poll(&mut events, None)?;
```

Comment:
Yeah my concern with adding a timeout is that effectively becomes a timeout for how long we allow the hook to run which we do not currently enforce in any of our hooks and I did not want to introduce a new concept of hook timeout and especially not only for windows. I think the worst that would happen here is the health check would stall but since they run in their own thread it would not interfere with the supervisor or other services in general.